File: ref/client/ContractDetails.java
/* Copyright (C) 2024 Interactive Brokers LLC. All rights reserved. This code is subject to the terms
 * and conditions of the IB API Non-Commercial License or the IB API Commercial License, as applicable. */

package com.ib.client;

import java.util.List;

import com.ib.client.Types.FundAssetType;
import com.ib.client.Types.FundDistributionPolicyIndicator;
import com.ib.client.Types.SecType;

public class ContractDetails {
    private Contract m_contract;
    private String   m_marketName;
    private double   m_minTick;
    private int      m_priceMagnifier;
    private String   m_orderTypes;
    private String   m_validExchanges;
    private int      m_underConid;
    private String   m_longName;
    private String   m_contractMonth;
    private String   m_industry;
    private String   m_category;
    private String   m_subcategory;
    private String   m_timeZoneId;
    private String   m_tradingHours;
    private String   m_liquidHours;
    private String   m_evRule;
    private double   m_evMultiplier;
    private List<TagValue> m_secIdList; // CUSIP/ISIN/etc.
    private int      m_aggGroup;
    private String   m_underSymbol;
    private String   m_underSecType;
    private String   m_marketRuleIds;
    private String   m_realExpirationDate;
    private String   m_lastTradeTime;
    private String   m_stockType; // COMMON, ETF, ADR etc.
    private Decimal  m_minSize;
    private Decimal  m_sizeIncrement;
    private Decimal  m_suggestedSizeIncrement;

    // BOND values
    private String   m_cusip;
    private String   m_ratings;
    private String   m_descAppend;
    private String   m_bondType;
    private String   m_couponType;
    private boolean  m_callable = false;
    private boolean  m_putable = false;
    private double   m_coupon = 0;
    private boolean  m_convertible = false;
    private String   m_maturity;
    private String   m_issueDate;
    private String   m_nextOptionDate;
    private String   m_nextOptionType;
    private boolean  m_nextOptionPartial = false;
    private String   m_notes;
    
    // FUND values
    private String   m_fundName;
    private String   m_fundFamily;
    private String   m_fundType;
    private String   m_fundFrontLoad;
    private String   m_fundBackLoad;
    private String   m_fundBackLoadTimeInterval;
    private String   m_fundManagementFee;
    private boolean  m_fundClosed = false;
    private boolean  m_fundClosedForNewInvestors = false;
    private boolean  m_fundClosedForNewMoney = false;
    private String   m_fundNotifyAmount;
    private String   m_fundMinimumInitialPurchase;
    private String   m_fundSubsequentMinimumPurchase;
    private String   m_fundBlueSkyStates;
    private String   m_fundBlueSkyTerritories;
    private FundDistributionPolicyIndicator m_fundDistributionPolicyIndicator;
    private FundAssetType m_fundAssetType;
    private List<IneligibilityReason> m_ineligibilityReasonList;

    // Get
    public int conid()                  { return m_contract.conid(); }
    public Contract contract()          { return m_contract; }
    public String marketName()          { return m_marketName; }
    public double minTick()             { return m_minTick; }
    public int priceMagnifier()         { return m_priceMagnifier; }
    public String orderTypes()          { return m_orderTypes; }
    public String validExchanges()      { return m_validExchanges; }
    public int underConid()             { return m_underConid; }
    public String longName()            { return m_longName; }
    public String contractMonth()       { return m_contractMonth; }
    public String industry()            { return m_industry; }
    public String category()            { return m_category; }
    public String subcategory()         { return m_subcategory; }
    public String timeZoneId()          { return m_timeZoneId; }
    public String tradingHours()        { return m_tradingHours; }
    public String liquidHours()         { return m_liquidHours; }
    public String evRule()              { return m_evRule; }
    public double evMultiplier()        { return m_evMultiplier; }
    public List<TagValue> secIdList()   { return m_secIdList; }
    public int aggGroup()               { return m_aggGroup; }
    public String underSymbol()         { return m_underSymbol; }
    public String underSecType()        { return m_underSecType; }
    public String marketRuleIds()       { return m_marketRuleIds; }
    public String realExpirationDate()  { return m_realExpirationDate; }
    public String lastTradeTime()       { return m_lastTradeTime; }
    public String stockType()           { return m_stockType; }
    public Decimal minSize()            { return m_minSize; }
    public Decimal sizeIncrement()      { return m_sizeIncrement; }
    public Decimal suggestedSizeIncrement() { return m_suggestedSizeIncrement; }
    
    public String cusip()               { return m_cusip; }
    public String ratings()             { return m_ratings; }
    public String descAppend()          { return m_descAppend; }
    public String bondType()            { return m_bondType; }
    public String couponType()          { return m_couponType; }
    public boolean callable()           { return m_callable; }
    public boolean putable()            { return m_putable; }
    public double coupon()              { return m_coupon; }
    public boolean convertible()        { return m_convertible; }
    public String maturity()            { return m_maturity; }
    public String issueDate()           { return m_issueDate; }
    public String nextOptionDate()      { return m_nextOptionDate; }
    public String nextOptionType()      { return m_nextOptionType; }
    public boolean nextOptionPartial()  { return m_nextOptionPartial; }
    public String notes()               { return m_notes; }

    public String fundName()                      { return m_fundName; }
    public String fundFamily()                    { return m_fundFamily; }
    public String fundType()                      { return m_fundType; }
    public String fundFrontLoad()                 { return m_fundFrontLoad; }
    public String fundBackLoad()                  { return m_fundBackLoad; }
    public String fundBackLoadTimeInterval()      { return m_fundBackLoadTimeInterval; }
    public String fundManagementFee()             { return m_fundManagementFee; }
    public boolean fundClosed()                   { return m_fundClosed; }
    public boolean fundClosedForNewInvestors()    { return m_fundClosedForNewInvestors; }
    public boolean fundClosedForNewMoney()        { return m_fundClosedForNewMoney; }
    public String fundNotifyAmount()              { return m_fundNotifyAmount; }
    public String fundMinimumInitialPurchase()    { return m_fundMinimumInitialPurchase; }
    public String fundSubsequentMinimumPurchase() { return m_fundSubsequentMinimumPurchase; }
    public String fundBlueSkyStates()             { return m_fundBlueSkyStates; }
    public String fundBlueSkyTerritories()        { return m_fundBlueSkyTerritories; }
    public FundDistributionPolicyIndicator fundDistributionPolicyIndicator() { return m_fundDistributionPolicyIndicator; }
    public FundAssetType fundAssetType()          { return m_fundAssetType; }
    public List<IneligibilityReason> ineligibilityReasonList() { return m_ineligibilityReasonList; }
    
    // Set
    public void contract(Contract contract)         { m_contract = contract; }
    public void marketName(String marketName)       { m_marketName = marketName; }
    public void minTick(double minTick)             { m_minTick = minTick; }
    public void priceMagnifier(int priceMagnifier)  { m_priceMagnifier = priceMagnifier; }
    public void orderTypes(String orderTypes)       { m_orderTypes = orderTypes; }
    public void validExchanges(String validExchanges) { m_validExchanges = validExchanges; }
    public void underConid(int underConid)          { m_underConid = underConid; }
    public void longName(String longName)           { m_longName = longName; }
    public void contractMonth(String contractMonth) { m_contractMonth = contractMonth; }
    public void industry(String industry)           { m_industry = industry; }
    public void category(String category)           { m_category = category; }
    public void subcategory(String subcategory)     { m_subcategory = subcategory; }    
    public void timeZoneId(String timeZoneId)       { m_timeZoneId = timeZoneId; }
    public void tradingHours(String tradingHours)   { m_tradingHours = tradingHours; }
    public void liquidHours(String liquidHours)     { m_liquidHours = liquidHours; }
    public void evRule(String evRule)               { m_evRule = evRule; }
    public void evMultiplier(double evMultiplier)   { m_evMultiplier = evMultiplier; }
    public void secIdList(List<TagValue> secIdList) { m_secIdList = secIdList; }
    public void aggGroup(int aggGroup)              { m_aggGroup = aggGroup; }
    public void underSymbol(String underSymbol)     { m_underSymbol = underSymbol; }
    public void underSecType(String underSecType)   { m_underSecType = underSecType; }
    public void marketRuleIds(String marketRuleIds) { m_marketRuleIds = marketRuleIds; }
    public void realExpirationDate(String realExpirationDate) { m_realExpirationDate = realExpirationDate; }
    public void  lastTradeTime(String lastTradeTime) { m_lastTradeTime = lastTradeTime; }
    public void stockType(String stockType)         { m_stockType = stockType; }
    public void minSize(Decimal minSize)            { m_minSize = minSize; }
    public void sizeIncrement(Decimal sizeIncrement) { m_sizeIncrement = sizeIncrement; }
    public void suggestedSizeIncrement(Decimal suggestedSizeIncrement) { m_suggestedSizeIncrement = suggestedSizeIncrement; }
    
    public void cusip(String cusip)             { m_cusip = cusip; }
    public void ratings(String ratings)         { m_ratings = ratings; }
    public void descAppend(String descAppend)   { m_descAppend = descAppend; }
    public void bondType(String bondType)       { m_bondType = bondType; }
    public void couponType(String couponType)   { m_couponType = couponType; }
    public void callable(boolean callable)      { m_callable = callable; }
    public void putable(boolean putable)        { m_putable = putable; }
    public void coupon(double coupon)           { m_coupon = coupon; }
    public void convertible(boolean convertible) { m_convertible = convertible; }
    public void maturity(String maturity)       { m_maturity = maturity; }
    public void issueDate(String issueDate)     { m_issueDate = issueDate; }
    public void nextOptionDate(String nextOptionDate) { m_nextOptionDate = nextOptionDate; }    
    public void nextOptionType(String nextOptionType) { m_nextOptionType = nextOptionType; }
    public void nextOptionPartial(boolean nextOptionPartial) { m_nextOptionPartial = nextOptionPartial; }
    public void notes(String notes)             { m_notes = notes; }

    public void fundName(String fundName)                                     { m_fundName = fundName; }
    public void fundFamily(String fundFamily)                                 { m_fundFamily = fundFamily; }
    public void fundType(String fundType)                                     { m_fundType = fundType; }
    public void fundFrontLoad(String fundFrontLoad)                           { m_fundFrontLoad = fundFrontLoad; }
    public void fundBackLoad(String fundBackLoad)                             { m_fundBackLoad = fundBackLoad; }
    public void fundBackLoadTimeInterval(String fundBackLoadTimeInterval)     { m_fundBackLoadTimeInterval = fundBackLoadTimeInterval; }
    public void fundManagementFee(String fundManagementFee)                   { m_fundManagementFee = fundManagementFee; }
    public void fundClosed(boolean fundClosed)                                { m_fundClosed = fundClosed; }
    public void fundClosedForNewInvestors(boolean fundClosedForNewInvestors)  { m_fundClosedForNewInvestors = fundClosedForNewInvestors; }
    public void fundClosedForNewMoney(boolean fundClosedForNewMoney)          { m_fundClosedForNewMoney = fundClosedForNewMoney; }
    public void fundNotifyAmount(String fundNotifyAmount)                     { m_fundNotifyAmount = fundNotifyAmount; }
    public void fundMinimumInitialPurchase(String fundMinimumInitialPurchase) { m_fundMinimumInitialPurchase = fundMinimumInitialPurchase; }
    public void fundSubsequentMinimumPurchase(String fundSubsequentMinimumPurchase) { m_fundSubsequentMinimumPurchase = fundSubsequentMinimumPurchase; }
    public void fundBlueSkyStates(String fundBlueSkyStates)                   { m_fundBlueSkyStates = fundBlueSkyStates; }
    public void fundBlueSkyTerritories(String fundBlueSkyTerritories)         { m_fundBlueSkyTerritories = fundBlueSkyTerritories; }
    public void fundDistributionPolicyIndicator(FundDistributionPolicyIndicator fundDistributionPolicyIndicator) { m_fundDistributionPolicyIndicator = fundDistributionPolicyIndicator; }
    public void fundAssetType(FundAssetType fundAssetType)                    { m_fundAssetType = fundAssetType; }
    public void ineligibilityReasonList(List<IneligibilityReason> ineligibilityReasonList) { m_ineligibilityReasonList = ineligibilityReasonList; }
    
    public ContractDetails() {
        m_contract = new Contract();
        m_minTick = 0;
        m_underConid = 0;
        m_evMultiplier = 0;
    }

    @Override public String toString() {
        StringBuilder sb = new StringBuilder( m_contract.toString() );

        add( sb, "marketName", m_marketName);
        add( sb, "minTick", m_minTick);
        add( sb, "priceMagnifier", m_priceMagnifier);
        add( sb, "orderTypes", m_orderTypes);
        add( sb, "validExchanges", m_validExchanges);
        add( sb, "underConId", m_underConid);
        add( sb, "longName", m_longName);
        add( sb, "contractMonth", m_contractMonth);
        add( sb, "industry", m_industry);
        add( sb, "category", m_category);
        add( sb, "subcategory", m_subcategory);
        add( sb, "timeZoneId", m_timeZoneId);
        add( sb, "tradingHours", m_tradingHours);
        add( sb, "liquidHours", m_liquidHours);
        add( sb, "evRule", m_evRule);
        add( sb, "evMultiplier", m_evMultiplier);
        add( sb, "aggGroup", m_aggGroup);
        add( sb, "underSymbol", m_underSymbol);
        add( sb, "underSecType", m_underSecType);
        add( sb, "marketRuleIds", m_marketRuleIds);
        add( sb, "realExpirationDate", m_realExpirationDate);
        add( sb, "lastTradeTime", m_lastTradeTime);
        add( sb, "stockType", m_stockType);
        add( sb, "minSize", m_minSize);
        add( sb, "sizeIncrement", m_sizeIncrement);
        add( sb, "suggestedSizeIncrement", m_suggestedSizeIncrement);

        add( sb, "cusip", m_cusip);
        add( sb, "ratings", m_ratings);
        add( sb, "descAppend", m_descAppend);
        add( sb, "bondType", m_bondType);
        add( sb, "couponType", m_couponType);
        add( sb, "callable", m_callable);
        add( sb, "putable", m_putable);
        add( sb, "coupon", m_coupon);
        add( sb, "convertible", m_convertible);
        add( sb, "maturity", m_maturity);
        add( sb, "issueDate", m_issueDate);
        add( sb, "nextOptionDate", m_nextOptionDate);
        add( sb, "nextOptionType", m_nextOptionType);
        add( sb, "nextOptionPartial", m_nextOptionPartial);
        add( sb, "notes", m_notes);

        if (m_contract.secType() == SecType.FUND) {
            add( sb, "fundName", m_fundName);
            add( sb, "fundFamily", m_fundFamily);
            add( sb, "fundType", m_fundType);
            add( sb, "fundFrontLoad", m_fundFrontLoad);
            add( sb, "fundBackLoad", m_fundBackLoad);
            add( sb, "fundBackLoadTimeInterval", m_fundBackLoadTimeInterval);
            add( sb, "fundManagementFee", m_fundManagementFee);
            add( sb, "fundClosed", m_fundClosed);
            add( sb, "fundClosedForNewInvestors", m_fundClosedForNewInvestors);
            add( sb, "fundClosedForNewMoney", m_fundClosedForNewMoney);
            add( sb, "fundNotifyAmount", m_fundNotifyAmount);
            add( sb, "fundMinimumInitialPurchase", m_fundMinimumInitialPurchase);
            add( sb, "fundSubsequentMinimumPurchase", m_fundSubsequentMinimumPurchase);
            add( sb, "fundBlueSkyStates", m_fundBlueSkyStates);
            add( sb, "fundBlueSkyTerritories", m_fundBlueSkyTerritories);
            add( sb, "fundDistributionPolicyIndicator", m_fundDistributionPolicyIndicator != null ? m_fundDistributionPolicyIndicator.name() : "");
            add( sb, "fundAssetType", m_fundAssetType != null ? m_fundAssetType.name() : "");
        }

        add( sb, "ineligibilityReasonList", EWrapperMsgGenerator.contractDetailsIneligibilityReasonList(this));

        return sb.toString();
    }

    public static void add(StringBuilder sb, String tag, Object val) {
        if (val == null || val instanceof String && ((String)val).length() == 0) {
            return;
        }
        sb.append( tag);
        sb.append( '\t');
        sb.append( val);
        sb.append( '\n');
    }
}

File: src/api/contract/contractDetails.ts
import TagValue from "../data/container/tag-value";
import SecType from "../data/enum/sec-type";
import { Contract } from "./contract";
import { FundAssetType, FundDistributionPolicyIndicator } from "./fund";
import { IneligibilityReason } from "./ineligibilityReason";

/**
 * Extended contract details.
 */
export interface ContractDetails {
  /** A fully-defined [[Contract]] object. */
  contract: Contract;

  /** The market name for this product. */
  marketName?: string;

  /**
   * The minimum allowed price variation.
   * Note that many securities vary their minimum tick size according to their price.
   *
   * This value will only show the smallest of the different minimum tick sizes regardless of the product's price.
   *
   * Full information about the minimum increment price structure can be obtained with the reqMarketRule function or the IB Contract and Security Search site.
   */
  minTick?: number;

  /**
   * Allows execution and strike prices to be reported consistently with market data, historical data and the order price, i.e. Z on LIFFE is reported in Index points and not GBP.
   *
   * In TWS versions prior to 972, the price magnifier is used in defining future option strike prices (e.g. in the API the strike is specified in dollars, but in TWS it is specified in cents).
   *
   * In TWS versions 972 and higher, the price magnifier is not used in defining futures option strike prices so they are consistent in TWS and the API.
   */
  priceMagnifier?: number;

  /** Supported order types for this product.  */
  orderTypes?: string;

  /**
   * Valid exchange fields when placing an order for this contract.
   *
   * The list of exchanges will is provided in the same order as the corresponding [[marketRuleIds]] list.
   */
  validExchanges?: string;

  /** For derivatives, the contract ID (conID) of the underlying instrument. */
  underConId?: number;

  /**	Descriptive name of the product. */
  longName?: string;

  /** Typically the contract month of the underlying for a Future contract. */
  contractMonth?: string;

  /** The industry classification of the underlying/product. For example, Financial. */
  industry?: string;

  /** The industry category of the underlying. For example, InvestmentSvc. */
  category?: string;

  /**	The industry subcategory of the underlying. For example, Brokerage. */
  subcategory?: string;

  /** The time zone for the trading hours of the product. For example, EST. */
  timeZoneId?: string;

  /**
   * The trading hours of the product.
   * This value will contain the trading hours of the current day as well as the next's.
   * For example, 20090507:0700-1830,1830-2330;20090508:CLOSED.
   *
   * In TWS versions 965+ there is an option in the Global Configuration API settings to return 1 month of trading hours.
   *
   * In TWS version 970+, the format includes the date of the closing time to clarify potential ambiguity, ex: 20180323:0400-20180323:2000;20180326:0400-20180326:2000.
   *
   * The trading hours will correspond to the hours for the product on the associated exchange.
   * The same instrument can have different hours on different exchanges.
   */
  tradingHours?: string;

  /**
   * The liquid hours of the product.
   * This value will contain the liquid hours (regular trading hours) of the contract on the specified exchange.
   *
   * Format for TWS versions until 969: 20090507:0700-1830,1830-2330;20090508:CLOSED.
   *
   * In TWS versions 965+ there is an option in the Global Configuration API settings to return 1 month of trading hours.
   *
   * In TWS v970 and above, the format includes the date of the closing time to clarify potential ambiguity, e.g. 20180323:0930-20180323:1600;20180326:0930-20180326:1600.
   */
  liquidHours?: string;

  /**
   * Contains the Economic Value Rule name and the respective optional argument.
   * The two values should be separated by a colon.
   *
   * For example, aussieBond:YearsToExpiration=3.
   *
   * When the optional argument is not present, the first value will be followed by a colon.
   */
  evRule?: string;

  /**
   * Tells you approximately how much the market value of a contract would change if the price were to change by 1.
   *
   * It cannot be used to get market value by multiplying the price by the approximate multiplier.
   */
  evMultiplier?: number;

  /**
   * Aggregated group Indicates the smart-routing group to which a contract belongs.
   *
   * Contracts which cannot be smart-routed have aggGroup = -1.
   */
  aggGroup?: number;

  /**
   * A list of contract identifiers that the customer is allowed to view CUSIP/ISIN/etc.
   *
   * For US stocks, receiving the ISIN requires the CUSIP market data subscription.
   *
   * For Bonds, the CUSIP or ISIN is input directly into the symbol field of the Contract class.
   */
  secIdList?: TagValue[];

  /** For derivatives, the symbol of the underlying contract. */
  underSymbol?: string;

  /**	For derivatives, returns the underlying security type. */
  underSecType?: SecType;

  /** The list of market rule IDs separated by comma Market rule IDs can be used to determine the minimum price increment at a given price. */
  marketRuleIds?: string;

  /**
   * Real expiration date.
   *
   * Requires TWS 968+ and API v973.04+.
   */
  realExpirationDate?: string;

  /** Last trade time. */
  lastTradeTime?: string;

  /** Stock type.  */
  stockType?: string;

  /**
   * The nine-character bond CUSIP. For Bonds only.
   *
   * Receiving CUSIPs requires a CUSIP market data subscription. */
  cusip?: string;

  /**
   * Identifies the credit rating of the issuer.
   *
   * This field is not currently available from the TWS API.
   *
   * For Bonds only.
   *
   * A higher credit rating generally indicates a less risky investment.
   * Bond ratings are from Moody's and S&P respectively.
   *
   *  Not currently implemented due to bond market data restrictions.
   */
  ratings?: string;

  /**
   * A description string containing further descriptive information about the bond.
   *
   * For Bonds only.
   */
  descAppend?: string;

  /** The type of bond, such as "CORP.". */
  bondType?: string;

  /**
   * The type of bond coupon.
   *
   * This field is currently not available from the TWS API.
   *
   * For Bonds only.
   */
  couponType?: string;

  /**
   * If `true`, the bond can be called by the issuer under certain conditions.
   *
   * This field is currently not available from the TWS API.
   *
   * For Bonds only.
   */
  callable?: boolean;

  /**
   * If `true`, the bond can be sold back to the issuer under certain conditions.
   *
   * This field is currently not available from the TWS API.
   *
   * For Bonds only.
   */
  putable?: boolean;

  /**
   * The interest rate used to calculate the amount you will receive in interest payments over the course of the year.
   *
   * This field is currently not available from the TWS API.
   *
   * For Bonds only.
   */
  coupon?: number;

  /**
   * Values are True or False.
   *
   * If `true`, the bond can be converted to stock under certain conditions.
   *
   * This field is currently not available from the TWS API.
   *
   * For Bonds only.
   */
  convertible?: boolean;

  /**
   * The date on which the issuer must repay the face value of the bond.
   *
   * This field is currently not available from the TWS API.
   *
   * For Bonds only.
   *
   * Not currently implemented due to bond market data restrictions.
   */
  maturity?: string;

  /**
   * The date the bond was issued.
   *
   * This field is currently not available from the TWS API.
   *
   * For Bonds only.
   *
   * Not currently implemented due to bond market data restrictions.
   */
  issueDate?: string;

  /**
   * Only if bond has embedded options.
   *
   * This field is currently not available from the TWS API.
   *
   * Refers to callable bonds and putable bonds.
   *
   * Available in TWS description window for bonds.
   */
  nextOptionDate?: string;

  /**
   * Type of embedded option.
   *
   * This field is currently not available from the TWS API.
   *
   * Only if bond has embedded options.
   */
  nextOptionType?: string;

  /**
   * Only if bond has embedded options.
   *
   * This field is currently not available from the TWS API.
   *
   * For Bonds only.
   */
  nextOptionPartial?: boolean;

  /**
   * If populated for the bond in IB's database.
   *
   * For Bonds only.
   */
  notes?: string;

  /**
   * Order's minimal size.
   */
  minSize?: number;

  /**
   * Order's size increment.
   */
  sizeIncrement?: number;

  /**
   * Order's suggested size increment.
   */
  suggestedSizeIncrement?: number;

  // FUND values
  fundName?: string;
  fundFamily?: string;
  fundType?: string;
  fundFrontLoad?: string;
  fundBackLoad?: string;
  fundBackLoadTimeInterval?: string;
  fundManagementFee?: string;
  fundClosed?: boolean;
  fundClosedForNewInvestors?: boolean;
  fundClosedForNewMoney?: boolean;
  fundNotifyAmount?: string;
  fundMinimumInitialPurchase?: string;
  fundSubsequentMinimumPurchase?: string;
  fundBlueSkyStates?: string;
  fundBlueSkyTerritories?: string;
  fundDistributionPolicyIndicator?: FundDistributionPolicyIndicator;
  fundAssetType?: FundAssetType;
  ineligibilityReasonList?: IneligibilityReason[];
}

export default ContractDetails;
